eddsa/eddsa.go

```diff
@@ -0,0 +1,163 @@
+package eddsa
+
+import (
+	"crypto/cipher"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/ed25519"
+	"github.com/dedis/crypto/random"
+)
+
+var suite = ed25519.NewAES128SHA256Ed25519(false)
+
+// EdDSA implements the EdDSA signature algorithm according to
+// the RFC https://tools.ietf.org/html/draft-josefsson-eddsa-ed25519-02
+type EdDSA struct {
+	seed   []byte
+	prefix []byte
+	// Secret being already hashed + bit tweaked
+	Secret abstract.Secret
+	// Public is the corresponding public key
+	Public abstract.Point
+}
+
+// NewEdDSAKey will return a freshly generated key pair to use for generating
+// EdDSA signatures.
+// If stream == nil, it will take the random.Stream.
+func NewEdDSA(stream cipher.Stream) *EdDSA {
+	if stream == nil {
+		stream = random.Stream
+	}
+	buffer := random.NonZeroBytes(32, stream)
+	scalar := sha512.Sum512(buffer)
+	scalar[0] &= 0xf8
+	scalar[31] &= 0x3f
+	scalar[31] |= 0x40
+
+	secret := suite.Secret().SetBytes(scalar[:32])
+	public := suite.Point().Mul(nil, secret)
+
+	return &EdDSA{buffer, scalar[32:], secret, public}
+}
+
+// Prefix returns the Prefix as being the right part of
+// the hashed seed
+func (e *EdDSA) Prefix() []byte {
+	c := make([]byte, len(e.seed))
+	copy(c, e.seed)
+	return c
+}
+
+// MarshalBinary will return the representation used by
+// the reference implementation of SUPERCOP ref10
+// Namely seed || Public
+func (e *EdDSA) MarshalBinary() ([]byte, error) {
+	pBuff, err := e.Public.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	eddsa := make([]byte, 64)
+	copy(eddsa, e.seed)
+	copy(eddsa[32:], pBuff)
+	return eddsa, nil
+}
+
+// EdDSASign will return a EdDSA signature using Ed25519. The secret must be
+// Picked() so it can use the prefix, the right part of the hashing done in
+// Pick.
+// Takes
+// - suite is the suite to use
+// - a secret to use to sign the message
+// - A corresponding public key
+// - msg message to sign
+// NOTE: Code taken from the Python implementation from the RFC
+// https://tools.ietf.org/html/draft-josefsson-eddsa-ed25519-02
+func (e *EdDSA) Sign(msg []byte) ([]byte, error) {
+	hash := sha512.New()
+	hash.Write(e.prefix)
+	hash.Write(msg)
+
+	// deterministic random secret and its commit
+	r := suite.Secret().SetBytes(hash.Sum(nil))
+	R := suite.Point().Mul(nil, r)
+
+	// challenge
+	// H( R || Public || Msg)
+	hash.Reset()
+	Rbuff, err := R.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	Abuff, err := e.Public.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	hash.Write(Rbuff)
+	hash.Write(Abuff)
+	hash.Write(msg)
+
+	h := suite.Secret().SetBytes(hash.Sum(nil))
+
+	// response
+	// s = r  h * secret
+	s := suite.Secret().Mul(e.Secret, h)
+	s.Add(r, s)
+
+	sBuff, err := s.MarshalBinary()
```

Comment:
No, exactly the opposite, as discussed above.  MarshalBinary() always returns a fixed-size slice, and I'm pretty sure is what you want here.